[GSoC 2022] Multipage Image Decoder API

ocpalo opened the pull request:
I am trying to finalize #17753 pr. To understand the requirements and see what we can do I also take a look at #6574. 

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

alalek commented:
OpenCV API is primary C++ based. However we need to think about Python, Java bindings - and support them too.

ocpalo commented:
@alalek I am planning to do Python and Java bindings after we finalize this API interface. Thanks for the reminder!

vpisarev commented on modules/imgcodecs/include/opencv2/imgcodecs.hpp:
```diff
@@ -43,6 +43,8 @@
 #ifndef OPENCV_IMGCODECS_HPP
 #define OPENCV_IMGCODECS_HPP
 
+#include <utility>
```
do we need this header in public header?

vpisarev commented on modules/imgcodecs/include/opencv2/imgcodecs.hpp:
```diff
@@ -322,6 +324,59 @@ CV_EXPORTS_W bool haveImageReader( const String& filename );
 CV_EXPORTS_W bool haveImageWriter( const String& filename );
 
+class CV_EXPORTS ImageCollection {
+    String m_filename;
```
please, use 'pimpl'. Hide everything non-public in a dedicated impl class

vpisarev commented on modules/imgcodecs/include/opencv2/imgcodecs.hpp:
```diff
@@ -322,6 +324,59 @@ CV_EXPORTS_W bool haveImageReader( const String& filename );
+
+public:
+
+    struct Iterator;
```
do we need forward declaration here?

vpisarev commented on modules/imgcodecs/include/opencv2/imgcodecs.hpp:
```diff
@@ -322,6 +324,59 @@ CV_EXPORTS_W bool haveImageReader( const String& filename );
+    CV_WRAP size_t nimages() const;
+    Mat operator[](int index);
+    CV_WRAP Mat at(int index);
+    CV_WRAP void free(int index);
```
does not just release the image at `index` or it also shifts all the subsequent images so that the sequence of images is still contiguous? If the latter, can we use another STL-like name (erase?)

vpisarev commented on modules/imgcodecs/include/opencv2/imgcodecs.hpp:
```diff
@@ -322,6 +324,59 @@ CV_EXPORTS_W bool haveImageReader( const String& filename );
+
+    CV_WRAP static ImageCollection fromMultiPageImage(const std::string& img, int flags);
+
+    struct Iterator
```
Please, hide the implementation details as well. Let's keep the headers as clean as possible

vpisarev commented:
@ocpalo, thank you! Please, rename PR so that the title contains [GSoC 2022] in the beginning.

sturkmen72 commented on modules/imgcodecs/include/opencv2/imgcodecs.hpp:
```diff
@@ -322,6 +322,42 @@ CV_EXPORTS_W bool haveImageReader( const String& filename );
+    ImageCollection(const String& filename, int flags);
+
+    void setup(const String& img, int flags);
+    CV_WRAP size_t size() const;
```
please remove CV_WRAP for now.

ocpalo commented:
I think I can remove the cv::Exception part from the API and provide error codes. This way, API would be more efficient.

ocpalo commented on modules/imgcodecs/src/loadsave.cpp:
```diff
@@ -1032,6 +1035,222 @@ bool haveImageWriter( const String& filename )
+    this->setup(filename, flags);
+}
+
+void ImageCollection::IImageCollection::setup(String const& filename, int flags) {
```
We need this because if user wants to do std::vector<ImageCollection>. Renaming to `init` might be better

ocpalo commented on modules/imgcodecs/src/loadsave.cpp:
```diff
@@ -1032,6 +1035,222 @@ bool haveImageWriter( const String& filename )
+    reinitDecoder();
+}
+
+size_t ImageCollection::IImageCollection::size() const { return m_size; }
```
This returns the number of pages in image. I think i should find a new, better name for this one. So we can return the current page Size(width, height).

ocpalo commented:
I want to say some things about my last commit. I changed constructor parameters. Used 1 parameter(struct ICParameters stands for ImageCollectionParameters). This way, C++20 users can do something like this

```cpp
ImageCollection collection({.filename = "a.tiff", .flags = IMREAD_ANYCOLOR});
```

Second thing is, operator++(int) was very hard to implement. Because in this API, we do not store any images inside Container. We let user to decode with `operator*`. I implemented operator*(int) as this way:

```cpp
ImageCollection::iterator ImageCollection::iterator::operator++(int) {
    iterator tmp(nullptr);
    ICParams parameters = m_ref->parameters();
    tmp.m_ref = new ImageCollection(parameters);
    tmp.m_ref->pImpl->init(parameters);
    for(int i = 0; i < m_curr; ++i) {
        ++(tmp);
    }
    ++(*this);
    return tmp;
}
```

I needed to create another ImageCollection and iterate until current index. Then return this new ImageCollection::iterator. Now i realized there is memory leak because there is no way to deallocate this new ImageCollection. It seems it is not possible to implement operator++(int). If there is a solution, please let me know :)

vpisarev commented on modules/imgcodecs/test/test_read_write.cpp:
```diff
@@ -303,4 +303,75 @@ TEST(Imgcodecs_Image, write_umat)
+            continue;
+        }
+        auto curr = i;
+        bool isEqual = (sum(prev != curr) == Scalar(0,0,0,0));
```
when all pages are the same, it's quite bad test. We need something different. I would create a multi-page image with different pages, then stored each page as individual file and then compared each page of multi-page image with the same page read as individual image, e.g.:
```
ImageCollection imcol("multi-page.tiff");
int i = 0;
for (auto& page : imcol) {
     double diff = norm(page, imread(format("multi-page-%d.tiff", i), NORM_INF);
     EXPECT_EQ(diff, 0.);
     i++;
}
```

ocpalo commented on modules/imgcodecs/src/loadsave.cpp:
```diff
@@ -1032,6 +923,187 @@ bool haveImageWriter( const String& filename )
+                                __LINE__);
+
+    size_t count = 1;
+    while(m_decoder->nextPage()) count++;
```
This `nextPage` call is required to find the ImageCollection size. About extracting all position dependent information, I need to pass `m_decoder` and `this` to the iterator in order to do that. So, i am not sure if this is a good idea because operator[] and at() also decodes pages. `this` is required for `operator*` decoding. So it can put pages inside the `m_vector`. I think we need to discuss this.

vpisarev commented on modules/imgcodecs/src/loadsave.cpp:
```diff
@@ -1032,6 +923,265 @@ bool haveImageWriter( const String& filename )
+    this->init(filename, flags);
+}
+
+void ImageCollection::Impl::init(String const& filename, int flags) {
```
init() should close the already opened collection, if any

ocpalo replied:
Good point, i missed that. Will fix it asap.

vpisarev commented on modules/imgcodecs/src/loadsave.cpp:
```diff
@@ -1032,6 +923,265 @@ bool haveImageWriter( const String& filename )
+
+ImageCollection::iterator& ImageCollection::iterator::operator++() {
+    if(m_pCollection->pImpl->currentIndex() == m_curr) {
+        m_pCollection->pImpl->advance();
```
do we need to call advance() here? Can operator [] take care of advancing internal pointer?

asmorkalov replied:
@ocpalo The question is still open. I suspect, that it's a bug.

ocpalo replied:
Vadim also suspected that this is a bug. But I do not think so. He requested a test to break this and I implemented it and it seems it works fine. I did not included in the comment because it was not well-written test. If you think a specific scenario that can break this, please tell me. So I can included in the unit tests and see if its broken or not.

The idea is, only iterators can advance the to next page that satisfy the requirement which is iterator internal counter is equal to current page index. Since iterators should not pass `end()`, So if this requirement is not satisfied, it does not call `nextPage()`. So file status is protected. Iterator count is incremented in any case, which is required.

I will write a detailed description of how that incrementing index works tonight.

vpisarev commented:
@ocpalo, overall, it looks good now. I put a couple of small comments.

vpisarev commented:
ok, I realized that init() reassigned m_decoder, so it should be closed automatically.

asmorkalov commented on modules/imgcodecs/test/test_read_write.cpp:
```diff
@@ -303,4 +303,142 @@ TEST(Imgcodecs_Image, write_umat)
+        double diff = cv::norm(i, imread(page_files[index]), NORM_INF);
+        EXPECT_EQ(0., diff);
+        ++index;
+    }
```
It makes sense to check index after the loop to ensure that all pages has been read.

asmorkalov commented on modules/imgcodecs/test/test_read_write.cpp:
```diff
@@ -303,4 +303,142 @@ TEST(Imgcodecs_Image, write_umat)
+        double diff = cv::norm(i, imread(page_files[index]), NORM_INF);
+        EXPECT_EQ(0., diff);
+        ++index;
+    }
```
The same here.

asmorkalov commented on modules/imgcodecs/test/test_read_write.cpp:
```diff
@@ -303,4 +303,142 @@ TEST(Imgcodecs_Image, write_umat)
+    // backward decoding -> 5,4,3,2,1,0
+    for(int i = (int)collection.size() - 1; i >= 0; --i)
+    {
+        double diff = cv::norm(collection[i], imread(page_files[i]), NORM_INF);
```
I propose to extract `imread` as separate line and check it it's result is not empty.

asmorkalov commented on modules/imgcodecs/test/test_read_write.cpp:
```diff
@@ -303,4 +303,142 @@ TEST(Imgcodecs_Image, write_umat)
+
+    ImageCollection collection(filename, IMREAD_ANYCOLOR);
+    // backward decoding -> 5,4,3,2,1,0
+    for(int i = (int)collection.size() - 1; i >= 0; --i)
```
Need `collection.size()` check against constant.

asmorkalov commented on modules/imgcodecs/test/test_read_write.cpp:
```diff
@@ -303,4 +303,142 @@ TEST(Imgcodecs_Image, write_umat)
+    int index = 0;
+    for(auto &i: collection)
+    {
+        double diff = cv::norm(i, imread(page_files[index]), NORM_INF);
```
I propose to extract imread as separate line and check it it's result is not empty.

asmorkalov commented on modules/imgcodecs/include/opencv2/imgcodecs.hpp:
```diff
@@ -332,6 +332,49 @@ CV_EXPORTS_W bool haveImageReader( const String& filename );
+to the collection of the images and iterate over the collection. Decode the necessary page with operator*.
+
+The performance of page decoding is O(1) if collection is increment sequentially. If the user wants to access random page,
+then the time complexity is O(n) because the collection has to be reinitialized every time in order to go correct page.
```
O(n) is not fully true. For TIFF `TIFFReadDirectory() and readHeader()` are called, but not full decoding pipeline.

vpisarev replied:
how about "Complexity is O(n) because the collection has to be reinitialized every time in order to go to the correct page. However, the intermediate pages are not decoded during the process, so typically it's quite fast"?

asmorkalov commented on modules/imgcodecs/include/opencv2/imgcodecs.hpp:
```diff
@@ -332,6 +332,49 @@ CV_EXPORTS_W bool haveImageReader( const String& filename );
+The performance of page decoding is O(1) if collection is increment sequentially. If the user wants to access random page,
+then the time complexity is O(n) because the collection has to be reinitialized every time in order to go correct page.
+This is required because multipage codecs does not support going backwards.
+After decoding the one page, it is stored inside the collection cache. Hence, trying to get Mat object from already decoded page is O(1).
```
I propose to make caching optional. It slows down some cases a bit, but may dramatically reduce memory consumption.

vpisarev replied:
@asmorkalov, this can be made optional very easily. Just call release/releaseCache() after you finished processing of each image. I think, it's quite simple and yet good enough workaround. I suggest not to make the API more complex at the moment

ocpalo replied:
Here is the example. Suppose the scenario is user wants to read images sequentially then discard them.

```cpp
ImageCollection collection("some.tiff");

int index = 0;
for(auto& page: collection) {
    // process page here
    collection.release(index);
    index++;
}
```

So the total memory usage here is actually `collection.size() * sizeof(cv::Mat (empty)) + sizeof(current page)`. If user release every image after processing,  there are no extra memory usage.

asmorkalov replied:
Ok

asmorkalov commented on modules/imgcodecs/src/loadsave.cpp:
```diff
@@ -1032,6 +991,265 @@ bool haveImageWriter( const String& filename )
+    size_t size() const;
+    Mat at(int index);
+    Mat& operator[](int index);
+    void release(int index);
```
I propose to rename the method to `ReleaseCache()`. Other OpenCV classes that contain `release()` method finalize the algorithm, but not clean cache. Object usage after `release()` is not expected.

ocpalo replied:
I inspired from cv::Mat. The intention of the both call is almost the same. But i am fine with the renaming of the method name.

ocpalo replied:
@asmorkalov you marked this as resolved. In the todays meeting, Vadim also agreed with you to rename the method name. I got confused why you marked this as resolved. I was going to rename it. Should I?

asmorkalov replied:
My fault. I resolved wrong discussion. Renamed method looks better, thanks!

asmorkalov reviewed (commented):
The solution looks very good for me! It's the next step for HEIF support and it's analogues.

vpisarev commented on modules/imgcodecs/src/loadsave.cpp:
```diff
@@ -1032,6 +991,265 @@ bool haveImageWriter( const String& filename )
+}
+
+Mat& ImageCollection::Impl::operator[](int index) {
+    if(m_pages[index].empty()) {
```
range is already checked at the external method `ImageCollection::at()` method

ocpalo replied:
There are 2 methods in the ImageCollection to read images. `.at()` method and `operator[]` operator. As in the STD containers, `.at` method does out-of-range access check but operator[] does not. There is no need to check here.

asmorkalov replied:
`at()` calls `operator[]`, but not vise versa. In case if index is -1 you'll get out of bound access for std::vector. For index, let say 1e+6 the next loop will be iterated 1e+6 times:  `for(int i = 0; i != index && advance(); ++i) {}`. I propose to move the check from `at()` to `operator[]` or have it in both methods. In `std::vector<>` is reasonable to get rid of range check for performance reason. Range check is heavier than actual work. But it's not true for image codecs. Also it could be a security issue.

vpisarev commented on modules/imgcodecs/src/loadsave.cpp:
```diff
@@ -1032,6 +991,265 @@ bool haveImageWriter( const String& filename )
+    Mat readData();
+    bool advance();
+    int currentIndex() const;
+    void reinit();
```
agree

ocpalo commented:
I made the fixes. Please take a look again @vpisarev @asmorkalov.

ALso, the Windows10-x64 fail is not related with my pull request. Please rerun that workflow only if you can. I need to re-run all workflows.

asmorkalov commented on modules/imgcodecs/test/test_read_write.cpp:
```diff
@@ -303,4 +303,152 @@ TEST(Imgcodecs_Image, write_umat)
+
+    ImageCollection collection(filename, IMREAD_ANYCOLOR);
+
+    int index = 0;
```
it make sense to use `size_t` for `index` No need to add type conversion to checks.

asmorkalov reviewed (commented):
I'll take a look on CI issue. Please ignore it for now.

ocpalo commented:
`at() calls operator[], but not vise versa. In case if index is -1 you'll get out of bound access for std::vector. For index, let say 1e+6 the next loop will be iterated 1e+6 times: for(int i = 0; i != index && advance(); ++i) {}. I propose to move the check from at() to operator[] or have it in both methods. In std::vector<> is reasonable to get rid of range check for performance reason. Range check is heavier than actual work. But it's not true for image codecs. Also it could be a security issue.`

@asmorkalov Regarding this comment, I have changed the std::vector `operator[]` with the std::vector `.at()` method. In this implementation, I pre-allocate vectors size. Now, In the `Impl` class, it does out of range access check in the both cases. I think that is good enough for range checks. 

Somehow i can not reply to your comment. That is why i wrote here as seperate comment. Thanks for the dealing with the CI issue!

asmorkalov commented on modules/imgcodecs/src/loadsave.cpp:
```diff
@@ -1032,6 +991,265 @@ bool haveImageWriter( const String& filename )
+}
+
+Mat& ImageCollection::Impl::operator[](int index) {
+    if(m_pages.at(index).empty()) {
```
OpenCV 4.x relies on C++11 as default C++ standard. `at` was added in C++20. Let's use manual range check + [] to presume compatibility. Details: https://en.cppreference.com/w/cpp/container/vector/at

ocpalo replied:
`at` method exists in C++11. In C++20, they added constexpr.

asmorkalov replied:
Sorry, my fault.

asmorkalov commented on modules/imgcodecs/src/loadsave.cpp:
```diff
@@ -1032,6 +991,265 @@ bool haveImageWriter( const String& filename )
+        throw cv::Exception(1, "Index out of bounds",
+                            String("ImageCollection::at"),
+                            __FILE__,
+                            __LINE__);
```
OpenCV uses `CV_Assert` for such checks.

asmorkalov commented on modules/imgcodecs/src/loadsave.cpp:
```diff
@@ -1032,6 +991,265 @@ bool haveImageWriter( const String& filename )
+        throw cv::Exception(1, "Index out of bounds",
+                            String("ImageCollection::at"),
+                            __FILE__,
+                            __LINE__);
```
The same CV_Assert();

ocpalo commented:
The current iterator api does not implements operator->. Should I? 

```cpp
struct CV_EXPORTS iterator {
     Mat* operator->();
     ...
}

Mat* ImageCollection::iterator::operator->() {
    CV_Assert(m_pCollection);
    return &m_pCollection->getImpl()->operator[](m_curr);
}
```

So one can use this to do operations on the cache Mat objects. And I realized that my `at()` and operator[]` methods returns copy of the Mat object. I think I should return `Mat&` to get rid of expensive copy costs in case of return value is reference to Mat.

@asmorkalov , Could you take a look at my last commit and tell me what do you think?

I have checked the [std::vector::at()](https://en.cppreference.com/w/cpp/container/vector/at) implementation. This also returns reference type.

ocpalo commented:
Further explanation about `operator++()` advance call. There are 2 variables named `m_curr` and `m_current`. Maybe I should rename them. 

* m_curr  -> Iterator API
* m_current -> ImageCollection API

When `operator++` is called, either way, we need to increment m_curr to understand that iterator points to which page. 
On the other hande, m_current can be modified only if m_current is equal to m_curr. Suppose we created a new `ImageCollection` and `Iterator`. So both `m_curr` and `m_current` is initialized to 0. If we advance with `operator++()`, both will increase its internal variables because they are equal. Now both variable are equal to `1`. Lets say we create another iterator, `iterB`, . This iterator `m_current` is initialized to 0. If we call `operator++` on iterator `iterB`, it will not advance the page (`m_current`) because it is not equal to `iterB`'s m_curr. It should not because `iterB` is behind of the first iterator. 

So what happens if `iterB` is tries to decode page 1(index 0) without calling `operator++()`. In this case, ImageCollection's m_decoder will be reinitialized to points to first page. Then it will advance the pages until `iterB`'s m_current is equal to m_curr. Then it decodes the page. In this way, I keep the `m_current` determined. Even thought, one iterator goes beyond of the scope, if other iterator tries to decode, the ImageCollection will be reinitialized and will advance until that index.

asmorkalov commented on modules/imgcodecs/include/opencv2/imgcodecs.hpp:
```diff
@@ -332,6 +332,51 @@ CV_EXPORTS_W bool haveImageReader( const String& filename );
+    void init(const String& img, int flags);
+    size_t size() const;
+    Mat& at(int index);
+    Mat& operator[](int index);
```
Reference as return value destroys the idea with cache. `resetCache` calls `release` for all images, including the image user has reference to. It's not what expected. Previous version relied on reference counting inside of `cv::Mat` and worked well. I propose to restore it back.

ocpalo replied:
Actually, `releaseCache` only calls `release` method on the given index page. For example, suppose you decode all pages into cache,

```cpp
ImageCollection col("filename.txt");
for(auto& i: coll){
     *i; // Decode all pages so they can go to cache

coll.releaseCache(2); // only release page 3 from the cache. 
auto mat = coll.at(1); // No decoding happens because page 2 is still at the cache
```

With returning reference, now it is user responsibility to not call `releaseCache` on the index that he is referring to. 

What do you think? And why do you think `releaseCache` releases for all images? I have never mentioned that and noone gave me feedback about that. Should I make that method to release for all images?

asmorkalov replied:
I rechecked all cases and it looks like the only dangerous case is something like this:
```
auto current_frame = decoder.at(i);
decoder.releaseCache();
```
releaseCache invalidates references (`auto` is `Mat&`), but it's expected behavior. In all other cases like `cv::Mat img = decoder.at(i);` `cv::Mat` copy constructor is called and user code does not suffer. So you are right! Let's go ahead with references then.

ocpalo replied:
In your example, `auto` is deduced to `Mat`. If user wants to get reference to Mat object, he needs `auto&`. Anyway, it is still safe. Thanks!

asmorkalov commented:
> Further explanation about operator++() advance call. There are 2 variables named m_curr and m_current. Maybe I should rename them.
> 
>     m_curr -> Iterator API
>     m_current -> ImageCollection API

Rename - it's good idea.

asmorkalov commented:
it makes sense to return const cv::Mat& in `at` and `operator[]` to not poison channels, size, etc on user side.

asmorkalov reviewed (approved):
:+1: Well done!

asmorkalov commented:
@ocpalo Please make operator[] and at() return values constant reference and squash commits. I'll merge the patch then.

ocpalo commented:
Ok. Give me an hour or two.

Thank you for the reviews btw.

ocpalo commented:
I have made the changes and pushed. Tbh, I do not understand why we force users to use `const&` if they want to reference it. 

If we keep `operator[]` and `at()` as before, the user can use it like this

```cpp
cv::Mat& firstPage = collection[0]; // 1
const cv::Mat& cFirstPage = collection[0]; //2
```

After the changes, the first one(1) would not compile. I understand the reason that it is protection for the user but we should not do that because since he is taking it by reference, it is his reponsibility to deal with this kind of stuff.

asmorkalov commented:
It's weak protection, but it covers at least:
- type, shape, channels change.
- unexpected cv::Mat realloc, if user tries to put that mat as output argument.
Mat content is still in danger, but out-of-bound access will not happen.

asmorkalov reviewed (approved):
:+1:

vpisarev commented:
@ocpalo, @asmorkalov, thank you for polishing it! I think, it can be merged now
